### PR TITLE
docs/website: Add APISIX edge authorization article

### PR DESCRIPTION
**Pull request by Yilialinn**
## Why

OPA users integrating policy decisions at an API gateway need more than an allow/deny example. They need a documented input contract, tested Rego, sanitized decision logs, clear bundle freshness boundaries, and observable failure behavior.

## What

This draft adds an OPA-focused article using Apache APISIX 3.17.0 and OPA 1.19.0 as the pinned integration example. The article covers:

- deny-by-default Rego and unit tests;
- explicit APISIX-to-OPA input fields;
- request correlation with masked OPA decision logs;
- bundle updates and conditional HTTP retrieval;
- timeout, unavailability, invalid JSON, and undefined-result cases.

The article does not claim configurable fail-open behavior, decision caching, exactly-once semantics, or unexecuted performance results.

This is intentionally a content-only PR. The article includes the key configuration and policy snippets, the validation method, and the recorded result matrix needed to understand the checks. The complete test fixtures and validation harness are retained by the author and are not added to the OPA repository.

## Validation

- Full integration harness rerun on 2026-07-31 with APISIX 3.17.0, OPA 1.19.0, and etcd 3.5.15; all scenarios passed
- The four OPA commands printed in the article were also executed verbatim in a clean temporary directory; checks, 8/8 tests, and bundle creation passed
- OPA formatting equivalent for 1.19.0: `opa fmt --fail --list`
- `opa check --strict`
- `opa test -v` — 8/8 tests passed
- `opa build`
- reader GET 200; reader POST 403; administrator GET/POST 200
- unknown identity, missing tenant, and malformed input denied
- OPA unavailable 403; a decision request sent to a deterministic slow fault endpoint with a 100 ms plugin timeout also returned 403
- invalid JSON/missing `result` 503
- request ID present in both sanitized logs; seeded credentials absent
- bundle `v1` to `v2` activated without an APISIX Route change; conditional retrieval returned 304
- Docusaurus production build
- dprint, markdownlint, and repository lint checks

## Author review and tooling disclosure

Generative AI tools assisted with drafting and validation orchestration. I reviewed the final article, technical claims, and staged diff and take responsibility for this contribution.

I reviewed the final diff, added the DCO `Signed-off-by`, and will respond to maintainer feedback personally.

### CODE REVIEW

File: docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+title: "Policy Enforcement at the API Edge: Designing Reliable OPA Authorization with Apache APISIX"
+authors: ["yilialin"]
+date: 2026-07-31
+slug: policy-enforcement-at-api-edge-apache-apisix
+---
+
+Putting an authorization check at an API gateway can make enforcement
+consistent, but a working `allow` rule is only the beginning. The integration
+also needs a stable input contract, deny-by-default policy, predictable failure
+behavior, safe audit logs, and a policy-distribution model whose freshness is
```
Comment (charlieegan3) [inline docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md:11]: Safe audit logs sounded a bit strange to me, perhaps just audit logs would be enough.

Or redacted perhaps? e.g. not including things like PII.

File: docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+understood.
+
+This article develops those properties with Open Policy Agent (OPA) 1.19.0 and
```
Comment (charlieegan3) [inline docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md:14]: Mostly OPA is remains compatible for a long time, if a reader looks at this in N months/years, might we scare them off with outdated versions when the content is still 99% valid and correct? 

I might just avoid referencing the versions here.

File: docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+Apache APISIX 3.17.0. The configuration and results below were validated with
+these pinned versions. APISIX is the example policy caller; the design questions
+apply to any gateway that asks OPA for a decision over HTTP.
+
+## Draw the enforcement boundary first
+
+The gateway is well placed to enforce facts visible at the edge: authenticated
```
Comment (charlieegan3) [inline docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md:21]:
```suggestion
The API gateway is well placed to enforce based on facts visible at the edge such as: authenticated
```

File: docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+caller identity, tenant, HTTP method, path, and coarse-grained permission.
+Services should still enforce invariants that depend on domain state, such as
```
Comment (charlieegan3) [inline docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md:23]:
```suggestion
Services should still enforce rules that depend on domain state, such as
```

File: docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+
+This article develops those properties with Open Policy Agent (OPA) 1.19.0 and
+Apache APISIX 3.17.0. The configuration and results below were validated with
+these pinned versions. APISIX is the example policy caller; the design questions
+apply to any gateway that asks OPA for a decision over HTTP.
+
+## Draw the enforcement boundary first
+
+The gateway is well placed to enforce facts visible at the edge: authenticated
+caller identity, tenant, HTTP method, path, and coarse-grained permission.
```
Comment (charlieegan3) [inline docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md:22]:
```suggestion
caller identity, tenant, HTTP method, path, and coarse-grained permissions.
```

File: docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+Services should still enforce invariants that depend on domain state, such as
+whether a user owns a particular invoice. The identity provider authenticates
+the caller, APISIX constructs a policy input, OPA evaluates policy, and the
```
Comment (charlieegan3) [inline docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md:25]: I feel like there is a missing step where the gw is allowing and denying based on the OPA response.

File: docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+  G -->|"input document"| O["OPA"]
+  O -->|"decision"| G
+  G -->|"allowed request"| U["Upstream service"]
+  B["Bundle server"] -->|"policy + data"| O
+  G -. "access log" .-> L["Log store"]
+  O -. "masked decision log" .-> L
+```
+
+This boundary keeps policy ownership clear. Authentication runs before the OPA
+plugin in this example. OPA decides whether the authenticated Consumer can perform
```
Comment (charlieegan3) [inline docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md:40]:
```suggestion
plugin in this example. OPA decides whether the authenticated Consumer (user) can perform
```

File: docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md [hunk 8]
```diff
@@ -0,0 +1,352 @@
...
+
+APISIX sends an `input` document containing `request` and `var`. The request
+includes the method, path, host, query parameters, and headers. Route, Service,
+and Consumer objects are optional and disabled by default through
+`with_route`, `with_service`, and `with_consumer`.
+
+The policy needs the authenticated Consumer username, so the Route enables only
+`with_consumer`:
+
+```json
```
Comment (charlieegan3) [inline docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md:53]: It's not 100% clear from the preceding text here that this is actually the configuration, not the input document to OPA.

I would make it clear that this is the config for the gw.

You can use title="example" in markdown blocks to give blocks names.

File: docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md [hunk 9]
```diff
@@ -0,0 +1,352 @@
...
+    "policy": "edge/authz/decision",
+    "timeout": 500,
+    "keepalive": true,
+    "with_consumer": true
+  }
+}
+```
+
+More context is not automatically better. Every additional object enlarges the
+policy interface and the data that may reach a decision log. During validation,
```
Comment (charlieegan3) [inline docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md:68]: At this point, some readers might not know what a decision log is.

File: docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md [hunk 10]
```diff
@@ -0,0 +1,352 @@
...
+object. Never assume that an object named "consumer" contains identity metadata
+only.
+
+The contract should also define normalization. In this run, APISIX represented
+the tenant header as `x-tenant`, while authentication added
+`X-Consumer-Username` and `X-Credential-Identifier`. Rego should not depend on
+accidental header spelling without a test fixture that matches the gateway's
+observed JSON. The same rule applies to paths, query parameters, and identity:
+choose one representation, capture a sanitized fixture, and review changes like
+an API schema change.
```
Comment (charlieegan3) [inline docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md:85]: Are there any tools we can share here to show that APISIX is mostly stable and for users to spot changes?

File: docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md [hunk 11]
```diff
@@ -0,0 +1,352 @@
...
+
+## Make the decision total and deny by default
```
Comment (charlieegan3) [inline docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md:87]: 'Total' might not be clear to all readers here.

File: docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md [hunk 12]
```diff
@@ -0,0 +1,352 @@
...
+
+The policy maps authenticated usernames to roles and requires the `acme`
+tenant. Readers may use GET; administrators may use GET or POST:
+
+```rego
+package edge.authz
+
+import rego.v1
```
Comment (charlieegan3) [inline docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md:95]: This is not needed

File: docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md [hunk 13]
```diff
@@ -0,0 +1,352 @@
...
+the tenant header as `x-tenant`, while authentication added
+`X-Consumer-Username` and `X-Credential-Identifier`. Rego should not depend on
+accidental header spelling without a test fixture that matches the gateway's
+observed JSON. The same rule applies to paths, query parameters, and identity:
+choose one representation, capture a sanitized fixture, and review changes like
+an API schema change.
+
+## Make the decision total and deny by default
+
+The policy maps authenticated usernames to roles and requires the `acme`
```
Comment (charlieegan3) [inline docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md:89]: I think before showing the policy it'd be good to have an example of the input JSON that is sent from the proxy, once we have an example of that, I can have a go at improving the policy below as well.

File: docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md [hunk 14]
```diff
@@ -0,0 +1,352 @@
...
+property is that `decision` is always defined and always contains `allow`.
+APISIX queries `data.edge.authz.decision` and expects OPA's response to contain
+`result`. An undefined rule is not the same as a defined denial.
+
+Tests should cover both intended permissions and malformed inputs. With OPA
+1.19.0, the validation run used the following non-writing format check, strict
+validation, test, and bundle-build commands against its policy directory:
+
+```shell
+opa fmt --fail --list policy/
```
Comment (charlieegan3) [inline docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md:148]: I think we can keep this focused on the policy logic, we can show the test command, but other than that I'm not sure that check and fmt are needed.

File: docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md [hunk 15]
```diff
@@ -0,0 +1,352 @@
...
+opa check --strict policy/
+opa test -v policy/
+opa build policy/authz.rego policy/mask.rego data.json -o edge-bundle.tar.gz
+```
+
+OPA 1.19.0 does not expose an `opa fmt --check` flag;
+`--fail --list` provides the non-writing, non-zero-on-diff check used here.
+The author's retained validation harness contains the policy tests and fixture
```
Comment (charlieegan3) [inline docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md:156]: Erm? is this left over from AI or something. 😬

File: docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md [hunk 16]
```diff
@@ -0,0 +1,352 @@
...
+
+Policy and data have different change patterns. The Rego module defines the
+permission model; bundle data assigns identities to roles. Both are reviewed and
+versioned, but changing data should not require editing the APISIX Route. This
+separation is what the later bundle-update test verifies.
+
+## Test failure behavior as part of the contract
+
+Authorization failures are operational behavior, not implementation detail. The
+APISIX 3.17.0 OPA plugin has three relevant branches:
```
Comment (charlieegan3) [inline docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md:174]:
```suggestion
Authorization failures are important behavior and must be tested carefully to ensure that we are not allowing unauthorized access. The
APISIX 3.17.0 OPA plugin has three main cases to cover for users:
```

File: docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md [hunk 17]
```diff
@@ -0,0 +1,352 @@
...
+Exercise the exported decision rather than testing helper rules alone. A
+refactor may change how roles or tenants are derived, but the exact document
+queried by APISIX must remain defined and retain the same meaning.
+
+Policy and data have different change patterns. The Rego module defines the
+permission model; bundle data assigns identities to roles. Both are reviewed and
+versioned, but changing data should not require editing the APISIX Route. This
+separation is what the later bundle-update test verifies.
+
+## Test failure behavior as part of the contract
```
Comment (charlieegan3) [inline docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md:171]: This section is quite verbose and hard to follow as it talks about different types of testing. I think there are two parts, the OPA rego tests, and integration tests that check things like timeouts etc.

File: docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md [hunk 18]
```diff
@@ -0,0 +1,352 @@
...
+
+That distinction should appear in alerts. A rise in normal 403 policy denials
+may indicate client behavior or an intentional policy change. Messages such as
+`failed to process OPA decision`, `invalid response body`, or
+`invalid OPA decision format` identify the integration path and need a
+different response. Retain those APISIX error lines alongside the client-facing
+status matrix so that the diagnosis does not infer root cause from an HTTP code
+alone.
+
+Fail-open may be appropriate for a narrowly scoped, low-risk endpoint, but it
```
Comment (charlieegan3) [inline docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md:202]: > Fail-open may be appropriate for a narrowly scoped, low-risk endpoint

I think something like this might be best to call out earlier, as a design rather than a testing concern.

File: docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md [hunk 19]
```diff
@@ -0,0 +1,352 @@
...
+is an architectural choice rather than a switch in this plugin. The tested
+plugin has no configurable fail-open option. If an organization chooses a
+different availability tradeoff, it should implement and threat-model that
+path explicitly instead of describing the current integration as fail-open.
+
+## Correlate logs without leaking credentials
+
+OPA decision logs are valuable because they record the queried path, input,
+result, bundle revision, metrics, and OPA `decision_id`. They can also record
+secrets. Enable decision logging only after defining `data.system.log.mask`:
```
Comment (charlieegan3) [inline docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md:212]:
```suggestion
result, bundle revision, metrics, and OPA `decision_id`. OPA instances should be configured to drop secrets before forwarding logs. Enable decision logging only after defining `data.system.log.mask` policy:
```

File: docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md [hunk 20]
```diff
@@ -0,0 +1,352 @@
...
+
+```rego
+package system.log
+
+import rego.v1
```
Comment (charlieegan3) [inline docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md:217]: No required.

File: docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md [hunk 21]
```diff
@@ -0,0 +1,352 @@
...
+
+## Treat the input as a versioned interface
+
+APISIX sends an `input` document containing `request` and `var`. The request
+includes the method, path, host, query parameters, and headers. Route, Service,
+and Consumer objects are optional and disabled by default through
+`with_route`, `with_service`, and `with_consumer`.
+
+The policy needs the authenticated Consumer username, so the Route enables only
+`with_consumer`:
```
Comment (charlieegan3) [inline docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md:51]: it'd be good have an example of what data is sent with with_consumer is set. vs not.

File: docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md [hunk 22]
```diff
@@ -0,0 +1,352 @@
...
+plugin upgrades.
+
+Masking is not a substitute for minimization. Removing `auth_conf` at log time
+protects the decision event, but OPA still received the field. If policy does not
+need Consumer context, disable `with_consumer`. If it does, consider whether a
+smaller authenticated identity value can be supplied through a trusted,
+gateway-controlled header. Also review the decision result: reasons and response
+headers can contain sensitive data just as inputs can.
+
+The correlation ID has its own trust boundary. A client-provided value is useful
```
Comment (charlieegan3) [inline docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md:251]: There is some mixed terminology here for correlation ID and request ID etc. I think best to use the APISIX terms as we do for consumer etc.

This final paragraph is a little light on actionable details. I am unsure it's really relevant to the integration with OPA.

File: docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md [hunk 23]
```diff
@@ -0,0 +1,352 @@
...
+for a deterministic validation run, but a production gateway should validate or
+generate it according to local rules and prevent ambiguous duplicates. The
+property that matters is that one non-secret identifier is recorded consistently
+on both sides.
+
+## Keep bundle freshness separate from decision caching
+
+To validate policy freshness, build two versioned revisions with `opa build`
+and serve them from a static HTTP endpoint. OPA polls that endpoint every one
+to two seconds:
```
Comment (charlieegan3) [inline docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md:261]: it might be that the defaults are ok for most users. Unsure we need to be as explicit.

File: docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md [hunk 24]
```diff
@@ -0,0 +1,352 @@
...
+
+```yaml
+services:
+  bundle:
+    url: http://bundle-server
```
Comment (charlieegan3) [inline docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md:266]:
```suggestion
    url: https://bundles.example.com/
```

File: docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md [hunk 25]
```diff
@@ -0,0 +1,352 @@
...
+gateway-controlled header. Also review the decision result: reasons and response
+headers can contain sensitive data just as inputs can.
+
+The correlation ID has its own trust boundary. A client-provided value is useful
+for a deterministic validation run, but a production gateway should validate or
+generate it according to local rules and prevent ambiguous duplicates. The
+property that matters is that one non-secret identifier is recorded consistently
+on both sides.
+
+## Keep bundle freshness separate from decision caching
```
Comment (charlieegan3) [inline docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md:257]: it's not clear to me what's meant by decision caching here.

File: docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md [hunk 26]
```diff
@@ -0,0 +1,352 @@
...
+    polling:
+      min_delay_seconds: 1
+      max_delay_seconds: 2
+```
+
+Revision `v1` gives the reader role read-only access. Revision `v2` changes the
+role data so the same authenticated reader can POST. The second decision
+becomes active one second after publication without changing the APISIX Route.
+That is an observation for this local configuration, not a propagation
+guarantee.
```
Comment (charlieegan3) [inline docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md:281]: I'd say this is too detailed for this post.

File: docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md [hunk 27]
```diff
@@ -0,0 +1,352 @@
...
+It is bundle HTTP caching. APISIX `keepalive`, meanwhile, reuses the connection
+to OPA. Neither mechanism caches an authorization decision. The changed bundle
+affects the next evaluated request even though APISIX keepalive remains enabled.
+
+For production, choose a polling window that balances freshness and service
+load, monitor bundle activation, retain the last known good policy where
+appropriate, and consider bundle signing. State the expected propagation
+window in operational documentation.
+
+Bundle rollout also needs a failure plan. Record the revision active on each OPA
```
Comment (charlieegan3) [inline docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md:294]: I feel like some of these topics are better kept for another post, this post is more the get started with the integration post than the day 2 operations version.

File: docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md [hunk 28]
```diff
@@ -0,0 +1,352 @@
...
+
+ETag behavior reduces unchanged transfers but does not make a fleet immediately
+consistent. Different OPA instances may poll at different moments within their
+configured windows. If a policy change must become effective everywhere before
+another action occurs, periodic polling alone is not a coordination protocol;
+the rollout process needs an explicit readiness condition.
+
+## Production checklist
+
+Before placing OPA authorization at an API edge:
```
Comment (charlieegan3) [inline docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md:307]:
```suggestion
Before placing OPA authorization at an API edge with APISIX:
```

File: docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md [hunk 29]
```diff
@@ -0,0 +1,352 @@
...
+
+- document the gateway-to-OPA input and decision schemas;
+- send only the Route, Service, Consumer, headers, and claims policy requires;
+- define a total decision with deny-by-default behavior;
+- run strict checks and positive, negative, missing-field, and malformed-input
```
Comment (charlieegan3) [inline docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md:312]: It'd be nice to call out Regal https://www.openpolicyagent.org/projects/regal rather than just OPA check, earlier too. This will catch many bugs before they're deployed.

File: docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md [hunk 30]
```diff
@@ -0,0 +1,352 @@
...
+  tests;
+- inject timeout, unavailable, invalid JSON, and undefined-decision failures;
+- distinguish policy denials from integration failures in dashboards and logs;
+- mask credentials using paths observed in real decision events;
+- correlate gateway and OPA logs with a non-secret request identifier;
+- monitor bundle revision and activation, and document the polling window;
+- keep bundle transfer caching, connection reuse, and decision caching as
+  separate concepts.
+
+The important result is not that a gateway can call OPA. It is that the call has
```
Comment (charlieegan3) [inline docs/blog/2026-07-31-policy-enforcement-at-api-edge-apache-apisix.md:322]: I can almost see the idea here, but think this needs to be rephrased.

It reads as a conclusion but it's not actually at the end of the post

**Review by charlieegan3 — CHANGES_REQUESTED** [review]
Hi @Yilialinn thanks for the draft here. I have left a few comments.

I think the main thing is that we need to be a little more focused on the flow of setting up the integration showing the main contact points that new users see.

That is, the APISIX config, the input JSON APISIX, the policy, the output and the logging config. Any more than that, and I think the main message is liable to getting lost.

**Comment by charlieegan3** [comment]
Hi there, I'm going to close this until we have time to work on getting it in together. Lmk when you have a chance to come back to it and I'll do another review. Drop me a message on slack if you have any questions.
